File: lesson-7/java/Truck.java
/*
 * This is the truck class
 *
 * @author Mohamad Tanbari
 * @version 1.0.0
 * @since 2024-11-13
 */

/**
 * The Truck class.
 */
public class Truck extends Vehicle {
    /**
     * The liscense plate orf the vehicle.
     */
    private String liscensePlate;

    /**
     * The color of the vehicle.
     */
    private String color;

    /**
     * The number of doors the vehicle has.
     */
    private final int numberOfDoors = 2;

    /**
     * The current speed of the vehicle.
     */
    private int speed;

    /**
     * Constructor.
     *
     * @param color - The color you want the vehicle to start with.
     * @param liscensePlate - The liscense plate you want the vehicle to have.
     * @param maxSpeed - The max speed the vehicle can reach.
     */
    public Truck(
        String color,
        String liscensePlate,
        int maxSpeed) {
        // Set the properties
        super(color, maxSpeed);
        this.liscensePlate = liscensePlate;
    }

    /**
     * The getter for the liscensePlate property.
     *
     * @return a string containing the liscense plate.
     */
    public String getLiscensePlate() {
        return this.liscensePlate;
    }

    /**
     * The setter for the liscensePlate property.
     *
     * @param newLiscensePlate - The new liscense plate you want to assign.
     */
    public void setLiscensePlate(String newLiscensePlate) {
        this.liscensePlate = newLiscensePlate;
    }

    /**
     * The status getter to show info about all properties of the vehicle.
     *
     * @return A string containing all the properties.
     */
    public String status() {
        final String status = "\n-> Speed: " + this.speed
                            + "\n-> Max Speed: " + this.maxSpeed
                            + "\n-> Number of Doors: " + this.numberOfDoors
                            + "\n-> Liscense Plate: " + this.liscensePlate
                            + "\n-> Color: " + this.color;

        return status;
    }

    /**
     * The provideAir method.
     *
     * @param airPressure
     */
    public void provideAir(int airPressure) {
      this.speed = this.speed - airPressure / 2;
    }
}

File: lesson-8/ts/Truck.ts
/*
 * The truck subclass extending Vehicle
 *
 * Author: Mohamad Tanbari
 * Version: 1.0.0
 * Since: 2024-11-12
 */

import Vehicle from './Vehicle.ts'

/**
 * The truck class
 */
export class Truck extends Vehicle {
  /**
   * The licensePlate property
   */
  private _liscensePlate: string

  /**
   * The amount of doors the vehicle has
   */
  private readonly _numberOfDoors: number = 2

  /**
   * Constructor
   *
   * @param {string} color - Color of the truck.
   * @param {number} maxSpeed - The max speed of the truck.
   * @param {string} liscensePlate - The license plate of the truck.
   */
  constructor (color: string,
    maxSpeed: number,
    liscensePlate: string,
    amountOfWheels: number) {
    super(color, maxSpeed, amountOfWheels)
    this._liscensePlate = liscensePlate
  }

  /**
   * The getter for the licensePlate property.
   *
   * @return {string} licensePlate - The current license plate of the vehicle.
   */
  public get licensePlate (): string {
    return this._liscensePlate
  }

  /**
   * The setter for the licensePlate property.
   *
   * @param {string} newLicensePlate - The new license plate you want to assign.
   */
  public set licensePlate (newLicensePlate: string) {
    this._liscensePlate = newLicensePlate
  }

  /**
   * The getter for the number of doors the truck has.
   *
   * @return {number} The number of doors the vehicle has.
   */
  public get numberOfDoors (): number {
    return this._numberOfDoors
  }

  /**
   * The provide air method to give the truck some air.
   *
   * @param {number} airPressure
   */
  public provideAir (airPressure: number): void {
    super.speed = super.speed - airPressure / 2
  }

  /**
   * This getter gives info on all properties of the vehicle.
   *
   * @return {string} A string containing all the property values.
   */
  public status (): string {
    const statusString = `-> Speed: ${super.speed}\n` +
                         `-> Max Speed: ${super.maxSpeed}\n` +
                         `-> Number of doors: ${this.numberOfDoors}\n` +
                         `-> Liscense Plate: ${this.licensePlate}\n` +
                         `-> Wheels: ${super.numberOfWheels}\n` +
                         `-> Color: ${super.colour}`

    return statusString
  }
}
